c/meterpreter/source/extensions/stdapi/server/sys/process/memory.c

```diff
@@ -339,3 +340,359 @@ DWORD request_sys_process_memory_unlock(Remote *remote, Packet *packet)
+	const size_t current_max_match_length = max_match_length;
+
+	dprintf("[MEM SEARCH] Getting handles & proc addresses");
+	const HMODULE kernel32_dll = GetModuleHandleA("kernel32.dll");
+	if (kernel32_dll == NULL) { dprintf("[MEM SEARCH] Could not get kernel32.dll handle"); result = ERROR_INVALID_HANDLE; goto done; }
+
+	const HMODULE ntdll_dll = GetModuleHandleA("ntdll.dll");
+	if (ntdll_dll == NULL) { dprintf("[MEM SEARCH] Could not get ntdll.dll handle"); result = ERROR_INVALID_HANDLE; goto done; }
+
+	const HANDLE get_proc_address = GetProcAddress(kernel32_dll, "GetProcAddress");
+	if (get_proc_address == NULL) { dprintf("[MEM SEARCH] Could not get GetProcAddress handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+	const GETPROCADDRESS GetProcAddress = (GETPROCADDRESS)get_proc_address;
+
+	const HANDLE virtual_query_ex = GetProcAddress(kernel32_dll, "VirtualQueryEx");
+	if (virtual_query_ex == NULL) { dprintf("[MEM SEARCH] Could not get VirtualQueryEx handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+
+	const HANDLE open_process = GetProcAddress(kernel32_dll, "OpenProcess");
+	if (open_process == NULL) { dprintf("[MEM SEARCH] Could not get OpenProcess handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+
+	const HANDLE close_handle = GetProcAddress(kernel32_dll, "CloseHandle");
+	if (close_handle == NULL) { dprintf("[MEM SEARCH] Could not get CloseHandle handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+
+	const HANDLE nt_read_virtual_memory = GetProcAddress(ntdll_dll, "NtReadVirtualMemory");
+	if (nt_read_virtual_memory == NULL) { dprintf("[MEM SEARCH] Could not get NtReadVirtualMemory handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+
+	const OPENPROCESS OpenProcess = (OPENPROCESS)open_process;
+	const CLOSEHANDLE CloseHandle = (CLOSEHANDLE)close_handle;
+	const VIRTUALQUERYEX VirtualQueryEx = (VIRTUALQUERYEX)virtual_query_ex;
+	const NTREADVIRTUALMEMORY NtReadVirtualMemory = (NTREADVIRTUALMEMORY)nt_read_virtual_memory;
+
+	const DWORD process_vm_read = 0x0010;
+	const DWORD process_query_information = 0x0400;
+	const DWORD wanted_process_perms = process_vm_read | process_query_information;
+
+	dprintf("[MEM SEARCH] Opening process");
+	const HANDLE process_handle = OpenProcess(wanted_process_perms, FALSE, pid);
+	if (process_handle == NULL) { dprintf("[MEM SEARCH] Could not get process handle"); result = ERROR_INVALID_HANDLE; goto done; }
+
+	MEMORY_BASIC_INFORMATION mem = { 0 };
+	const size_t megabytes_64 = 64 * 1024 * 1024;
+
+	dprintf("[MEM SEARCH] Allocating buffer for storing process memory");
+	buffer = (char*)malloc(megabytes_64);
+	if (buffer == NULL) { dprintf("[MEM SEARCH] Could not allocate memory buffer"); result = ERROR_OUTOFMEMORY; goto done; }
+
+	// The maximum length of data that we can read into a buffer at a time from a memory region.
+	const size_t current_max_size = megabytes_64;

```

Comment:
Is this an arbitrary value, or does it come from somewhere as a recommended value, i.e. Microsoft docs etc?

Reply:
Its an arbitrary value. Microsoft doesn't mention any recommended values here, and I thought 64 MB was a suitable value. I'm not opposed to reducing this value 🚀